### PR TITLE
[FLINK-8306] [kafka] Introduce KafkaOffsetCommitter interface

### PR DESCRIPTION
**Pull request by tzulitai**
## What is the purpose of the change

This PR is built upon the reworked `FlinkKafkaConsumerBaseTest` in #5188.

Prior to this PR, offset committing was coupled tightly with the `AbstractFetcher`, making unit tests for offset committing behaviours hard to compose concisely. For example, we had tests that required mocking the offset commit methods on `AbstractFetcher`, while ideally, it would be best that those methods are made final (thus, unable to be mocked) to prevent accidental overrides.

This PR decouples offset committing as a separate service behind a new `KafkaOffsetCommitter` interface. For now, the `AbstractFetcher` is reused as an implementation of this service, so that this PR does not introduce any more change other than introducing a new layer of abstraction.

Instead of using mocks on `AbstractFetcher`, unit tests that verify offset committing behaviour now provide a dummy verifiable implementation of the `KafkaOffsetCommitter`, and test against that.

## Brief change log

- Migrate `AbstractFetcher::commitInternalOffsetsToKafka` method to the newly introduced `KafkaOffsetCommitter` interface.
- Let `AbstractFetcher` implement `KafkaOffsetCommitter`
- In the `FlinkKafkaConsumerBase`, let "offset committing" and "record fetching" be logically separated to be handled by two services, i.e. namely a `KafkaOffsetCommitter` and a `AbstractFetcher`. Physically, the fetcher instance sits behind both service abstractions.
- In `FlinkKafkaConsumerBaseTest`, remove all mocks on `AbstractFetcher::commitInternalOffsetsToKafka`, and test against a `KafkaOffsetCommitter` instead.
- Additional hotfix 2906968 that fixes a stale comment referring to an old `AbstractFetcher` behaviour to avoid confusion.

## Verifying this change

This PR does not add any new functionality. Reworked test also do not affect test coverage.
`FlinkKafkaConsumerBaseTest` verifies all changes.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? n/a

### CODE REVIEW

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java [hunk 1]
```diff
@@ -51,7 +50,7 @@
  *            the Flink data streams.
  * @param <KPH> The type of topic/partition identifier used by Kafka in the specific version.
  */
-public abstract class AbstractFetcher<T, KPH> {
+public abstract class AbstractFetcher<T, KPH> implements KafkaOffsetCommitter {
```
Comment (pnowojski) [inline flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java]: Maybe use composition instead of inheritance here? `AbstractFetcher` as "being a" `KafkaOffsetCommitter` do not make sense to me when I say it out laud, while `AbstractFetcher` "having a"  `KafkaOffsetCommitter` sounds better.

File: flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseTest.java [hunk 2]
```diff
@@ -89,30 +90,44 @@
...
 
 		final AssignerWithPeriodicWatermarks<String> periodicAssigner = mock(AssignerWithPeriodicWatermarks.class);
 		final AssignerWithPunctuatedWatermarks<String> punctuatedAssigner = mock(AssignerWithPunctuatedWatermarks.class);
 
-		DummyFlinkKafkaConsumer<String> c1 =
-			new DummyFlinkKafkaConsumer<>(mock(AbstractFetcher.class), mock(AbstractPartitionDiscoverer.class), false);
+		DummyFlinkKafkaConsumer<String> c1 = new DummyFlinkKafkaConsumer<>(
+				mock(AbstractFetcher.class),
+				mock(AbstractPartitionDiscoverer.class),
+				mock(KafkaOffsetCommitter.class),
```
Comment (pnowojski) [inline flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseTest.java]: Please replace everywhere `mock(KafkaOffsetCommitter.class)` with `new DummyKafkaOffsetCommitter()` or `new NoOpKafkaOffsetCommitter()`.

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java [hunk 3]
```diff
@@ -770,8 +777,8 @@ public final void notifyCheckpointComplete(long checkpointId) throws Exception {
 			return;
 		}
 
-		final AbstractFetcher<?, ?> fetcher = this.kafkaFetcher;
-		if (fetcher == null) {
+		final KafkaOffsetCommitter offsetCommitter = this.kafkaOffsetCommitter;
```
Comment (pnowojski) [inline flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java]: why do we need this `final` local variable? `kafkaOffsetCommitter` seems to be "write-once" variable.

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java [hunk 4]
```diff
@@ -559,6 +565,7 @@ public void onException(Throwable cause) {
 		//            the fetchers 'snapshotCurrentState()' method return at least
 		//            the restored offsets
 		this.kafkaFetcher = fetcher;
+		this.kafkaOffsetCommitter = createOffsetCommitter();
```
Comment (pnowojski) [inline flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java]: Changing order of those lines:
```
		this.kafkaFetcher = fetcher;
		this.kafkaOffsetCommitter = createOffsetCommitter();
```
will brake the code. This is super fragile and is kind of hard to follow.

**Comment by pnowojski** [comment]
Btw, doesn't the `org.apache.flink.streaming.connectors.kafka.internal.KafkaConsumerThread#setOffsetsToCommit` have a race condition on those lines:


```
		// record the work to be committed by the main consumer thread and make sure the consumer notices that
		if (nextOffsetsToCommit.getAndSet(offsetsToCommit) != null) {
			log.warn("Committing offsets to Kafka takes longer than the checkpoint interval. " +
					"Skipping commit of previous offsets because newer complete checkpoint offsets are available. " +
					"This does not compromise Flink's checkpoint integrity.");
		}
		this.offsetCommitCallback = commitCallback;
```

(`getAndSet` followed by `volatile`) with `org.apache.flink.streaming.connectors.kafka.internal.KafkaConsumerThread#run`? However currently it doesn't seem to be so important, since:

1. it's used only for metrics
2. and only first `successfulCommits.inc()` can be omitted because of that., because (at least now) `offsetCommitCallback` is a `write-once` variable, and any subsequent overwrites, are overwriting it to the same value.

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java (hunk 1 above)
Comment (tzulitai) [reply flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java]: I agree that composition suits better here, or maybe even neither of both. However, the reality is that currently the offset committing logic is implemented tightly as part of the `AbstractFetcher`, sharing the same Kafka client for both fetching records and committing offsets. Decoupling that would require further refactoring, which I think is a bit out of scope for the current issue at hand.

I have been thinking that we should simply have two separate service implementations for offset committing and record fetching. If that happens, then neither composition or inheritance is required; offset committing and record fetching simply lives as two separate services.

What do you think?

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java (hunk 4 above)
Comment (tzulitai) [reply flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java]: This is a very valid argument. Will address this with a factory perhaps, as you suggested.

File: flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseTest.java (hunk 2 above)
Comment (tzulitai) [reply flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBaseTest.java]: Will do.

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java (hunk 1 above)
Comment (tzulitai) [reply flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java]: The result could also very well be that we should use this opportunity to refactor the vague dependencies between fetcher / consumer thread / consumer base, and include in this PR. I would not be against that.

**Comment by tzulitai** [comment]
@pnowojski thanks a lot for your insightful review!

regarding the choice of composition or inheritance: actually, in the end I think we should be leaning towards neither of both, and let offset committing and record fetching live as separate components. I've left more detailed comments inline.

**Comment by tzulitai** [comment]
Regarding the race condition you mentioned:
hmm, I can't seem to exactly nail down the "and only first `successfulCommits.inc()` can be omitted because of that" case you mentioned, could you elaborate on that a bit more?

But yes, it seems like there certainly is a race condition here, and can even cause an NPE on:
https://github.com/apache/flink/blob/master/flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaConsumerThread.java#L496

Seems like the `nextOffsetsToCommit` and its callback instance should be bundled as a single `AtomicReference` here ...

Would you like to open a PR for that? I wanted to ask because you discovered it in the first place; if not I'll open a fix for it.

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java (hunk 1 above)
Comment (pnowojski) [reply flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java]: Probably you are right, however I haven't thought this through, but if you had and you feel that's a good opportunity to make some larger refactor I'm fine with that :)

Anyway, as we discussed offline, such refactor could be made in separate PR and here you could just implement `AbstractFetcher#doCommitInternalOffsetsToKafka` for the testing purposes. 

I don't mind if you choose one over the other approach :)

**Comment by pnowojski** [comment]
Regarding the race condition. I think we meant the same thing or maybe I misunderstood part of the code, but yes, this NPE seems like a possible outcome of this race condition.

If you have spare cycles to fix this race feel free to do so, I would like to avoid working on it in this and the following week because of some code handing over before some lengthy vacation in my team.

**Comment by tzulitai** [comment]
As discussed offline:
Will re-open a PR for this that avoids introducing a new interface. I agree that it would make more sense to introduce new interface when we actually decide to go for a major refactor ;)

**Comment by tzulitai** [comment]
Subsumed be re-opened PR: #5284
